manual/manual/refman/exten.etex

```diff
@@ -2461,4 +2461,117 @@ val x: int
   [@@@ocaml.alert deprecated "Please do something else"]
 \end{verbatim}
 
+\section{Generalized open statements\label{s:generalized-open}}
+
+(Introduced in 4.08)
+
+\begin{syntax}
+definition:
+      ...
+   |  'open'  module-expr
+   |  'open!' module-expr
+;
+specification:
+      ...
+   |  'open'  extended-module-path
+   |  'open!' extended-module-path
+;
+expr:
+       ...
+     | 'let' 'open'  module-expr 'in' expr
+     | 'let' 'open!' module-expr 'in' expr
+;
+\end{syntax}
+
+
+This extension makes it possible to open any module expression in
+module structures and expressions. A similar mechanism is also available
+inside module types, but only for extended module paths (e.g. "F(X).G(Y)").
+
+For instance, a module can be constrained when opened with
+
+\begin{caml_example*}{verbatim}[error]
+module M = struct let x = 0 let hidden = 1 end
+open (M:sig val x: int end)
+let y = hidden
+\end{caml_example*}
+
+
+Another possibility is to immediately open the result of a functor application
+
+\begin{caml_example}{verbatim}
+  let sort (type x) (x:x list) =
+    let open Set.Make(struct type t = x let compare=compare end) in
+    elements (of_list x)
+\end{caml_example}
+
+Going further, this construction can introduce private components inside a
+structure,
+
+\begin{caml_example}{verbatim}
+module M = struct
+  let x = 0
+  open! struct
+    let x = 0
+    let y = 1
+  end
+  let w = x + y
+end
+\end{caml_example}
+
+One important limitation of those extended opens is that they should not generate
+new unnamed types that are visible from the outside world:
```

Comment:
I don't really like that formulation, the types aren't exactly "unnamed", it's just that they can't be referred to from from the outside.
How about this instead:

> One important restriction is that types introduced by `open struct ... end` cannot appear in the signature of the structure containing the open, unless they are already available through some other name.
> So:
> ```ocaml
> module M = struct
>   open struct type 'a t = 'a option = None | Some of 'a end
>  let x : int t = Some 1
> end
> ```
> is OK, because `x` can be given the type `int option`, but:
> ```ocaml
> module M = struct
>   open struct type t = A end
>   let x = A
> end
> ```
> is not because `x` cannot be given any type other than `t`, which only exists locally. Although the above would be OK if `x` too was local:
> ```ocaml
> module M = struct
>   open struct type t = A end
>   [@@@ellipsis]
>   open struct let x = A end
>   [@@@ellipsis]
> end
> ```

Reply:
This is indeed a better formulation. However, the phrasing `signature of the structure containing the open` might be a a bit heavy. Maybe `signature of the parent/enclosing structure`?
As a minor nitpick, I would personally remove the colon `:` just after `So` since it feels redundant.

Reply:
I'm not sure how much it matters, but even this formulation is a bit misleading. It implies that:

```ocaml
module M = struct
  open struct type t = A end
  type s = t
  let x = A
end
```

would be allowed, but it isn't. Maybe talking in terms of the local types being "aliased to" or "defined equal to" some non-local type would be clearer.

Although, even that formulation is overly conservative, since something like:

```ocaml
module M = struct
  open struct type t = A end
  type s = t
  let x : s = A
end
```

is allowed.  However, I don't think we should encourage people to write things like this. Depending on how they're written they are dubious at best and non-principal at worst.